keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+  types, so the implementation that reads ValidatingAdmissionPolicy
+  instances and enforces them would have to live in k8s.io/apiserver and be
+  active in aggregated API servers to enforce admission on aggregated types
+  effectively (same as admission webhooks today).
+
+Plan:
+
+- Do not offer type checking for aggregated types.
+- Support ValidatingAdmissionPolicy in aggregated API servers.
+
+#### Metrics
+
+Goals:
+
+- Parity with admission webhook metrics
+  - Should include counter of deny, warn and audit violations
+- Counters for number of policy defintions and policy configurations in cluster
+  - Label by state (active vs. error), enforcement action (deny, warn)
+
+
+- xref: [Metrics Provided by OPA Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/metrics/)
+- xref: [Admission Webhook Metrics](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#admission-webhook-metrics)
+
+### User Stories
+
+In addition to "User Stores", see below "Potential Applications" for a list of
+known applications and their use case requirements.
+
+#### Use Case: Standalone Policy
+
+User wishes to define a simple policy that required no configuration. They don't
+want to create configuration CRD since what they're doing can be expressed quite
+simply in a single CEL expression.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  singletonPolicy: true
+  match:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+  - expression: "object.spec.replicas < 100"
+```
+
+#### Use Case: Shared Configuration
+
+User wishes to define a configuration for a list of banned words that may not be
+used in any of a wide range of identifiers in the cluster (resource names,
+container names, ...).
+
+- Configuration CRD is defined to hold the list of banned words.
+- Multiple policies are defined for different resources. The policies all
+  reference the same configuration CRD. 
+- Policy must be able to specify it's own matching rules since the configuration
+  applies cluster wide.
+- A single custom resource is defined with the list of banned words (but has not
+  matching rules of its own).
+
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: BannedWords
+metadata:
+  name: "banned-words.example.com"
+spec:
+  bannedWords:
+  - glitter
+  - rainbow
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy1.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.name in config.bannedWords"
```

Comment:
Do we need to consider resource constraints/failure modes for (potentially) relatively resource-heavy CEL expressions like this one?  At a high level, I feel like what was done for [CRD validation rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#resource-constraints) would be applicable (estimate how expensive the rules are to run before we create/apply and fail if it's too much).

I think we could have:
* Failure due to an individual validation rule exceeding a per-CEL-rule estimated cost limit
* Failure due to the sum total of a `ValidatingAdmissionPolicy`'s rules exceeding the per-policy estimated cost limit (if we decided to have one - I think it makes sense for the same reasons as CRD validation rules, to cap resource usage on a single policy)
* Failure due to a request timeout while evaluating a rule
 
That said, cost estimation [relies on type checking](https://github.com/google/cel-go/blob/1b994e0945736cb8ef2b51aedf4b30fdeeb7bc1c/checker/cost.go#L284), so maybe that'd mean cost limits would only be available some of the time if policy rules aren't always type-checked.

Reply:
Thanks for bringing this up.

I agree that estimated cost is, unfortunately, not something we can offer with any kind of guarantees attached due to the already listed issues that have prevented use from enforcing type safety. We could compute it for the same cases where we provide informational type checking, in which case we could report any cost limit violations in the same way we report type checking violations. To make matters worse, we will be handling both built-in and CRD types, and built-ins do not have their `max{Length,Properties,Items}` limits documented in schemas, which further limits estimated cost usefulness. I'll make note of this in the KEP.

Runtime cost limits are something that we should do, as is a context cancellation backstop. Both have `FailurePolicy` implications. I'll add this into phase 2.

Reply:
Shift-left checks for CEL estimated cost would be really helpful here. Something to think about when working on the CEL expression linter?